### ISSUE TITLE
Make some dependencies (e.g. keras) optional

### ISSUE THREAD

**Issue by benjamin-work** [RasaHQ/rasa_core#83]
As I see it, some components are optional for RASA core to work. For instance, if I don't want to use the `KerasPolicy`, I wouldn't need keras and hence tensorflow. Is it possible to make these requirements optional? This would really facilitate deployment.

**Comment by tmbo** [comment]
Yes, it is possible (using optional dependencies in the `setup.py`). But as long as there is not a good alternative (e.g. a policy using sklearn or something) I am not sure that is worth the additional care users need to take when installing (as they need to do something like `pip install rasa_core[keras]` - or whatever the syntax is).What is the use case for deploying it without Keras?

**Comment by benjamin-work** [comment]
That was exactly the use case: For our bot, an sklearn LogisticRegression was more than enough, the keras policy was overkill. And since tensorflow is a big dependency, it would be nice if we could skip it like you mentioned above.

If we provide an sklearn policy PR, would that be a good reason to make keras optional? (Although that would add scikit-learn as a new dependency.)

**Comment by tmbo** [comment]
Yes, that would be a very valid reason to make both, sklearn as well as keras optional dependencies 🙂 .

**Comment by benjamin-work** [comment]
Okay, I will see to that. I may have one or two questions along the way.

**Comment by tmbo** [comment]
sure, fire away when you get there :+1:

**Comment by benjamin-work** [comment]
First draft here:

https://github.com/benjamin-work/rasa_core/blob/sklearn-policy/rasa_core/policies/sklearn_policy.py

At the moment, the `SklearnPolicy` has support for custom cross-validation and grid search. When I get the tests running, I will add some unit tests.

There was some confusion because it seems that the target never contains a class 1. Instead, for N intents, there are the classes 0, 2, 3, ..., N+1, where 0 was the injected as "listen" class. For that reason, and because sklearn automatically infers the classes from the target, we had to insert a dummy class ([link](https://github.com/benjamin-work/rasa_core/blob/sklearn-policy/rasa_core/policies/sklearn_policy.py#L95)).

**Comment by tmbo** [comment]
@benjamin-work The domain has two default actions at the moment, listen and restart:

https://github.com/RasaHQ/rasa_core/blob/6fbbda837d4578564c77218f87480d50dfa9fbb7/rasa_core/domain.py#L90

I think (we encountered something similar in NLU) that sklearn can not handle the case where you declare a label to be present that was never encountered in the training data (and the restart action usually isn't part of the training data). So the number `1` you are missing is `ActionRestart`.

**Comment by benjamin-work** [comment]
So the correct solution would be to insert the missing class only in case that "restart" action was not part of the training set. What is the best way to find out that information?

Apart from that, does this look like the right direction?

**Comment by tmbo** [comment]
Yes, it is going in the right direction - good. job :+1:

I think you can get around the missing entry in the resulting probability array by explicitly using a LabelEncoder from sklearn.

**Comment by benjamin-work** [comment]
I made some adjustments and wrote a couple of tests:

* https://github.com/benjamin-work/rasa_core/blob/sklearn-policy/rasa_core/policies/sklearn_policy.py
* https://github.com/benjamin-work/rasa_core/blob/sklearn-policy/tests/test_policies.py#L125

Giving the user the choice of installing keras/sklearn is not part of the changes yet. Should it be or is it separate? If the latter, I would make the PR.

**Comment by tmbo** [comment]
Mhm so I think making `keras` optional can be separate, but I don't think it is a good idea to merge the PR making sklearn and keras required / listed dependencies - so at least the sklearn ones should be optional.

**Comment by benjamin-work** [comment]
Okay, but did you see my other comment that `run_story_evaluation` already has a dependency on sklearn? Should it still be optional?

**Comment by tmbo** [comment]
That is a very good point - let's add the policy first then and split up the dependencies afterwards.

**Comment by benjamin-work** [comment]
Great, I added the PR. I leave this open for the initial question.
